### ISSUE TITLE
Support DynamoDb as storage backend

### ISSUE THREAD

**Issue by john-tipper** [grafeas/grafeas#380]
**Is your feature request related to a problem? Please describe.**
I'd like to be able to use DynamoDB as a storage backend.  I don't want to pay for an always-on database and for this database to be highly-available.  I'd like to use a cloud-native solution.

**Describe the solution you'd like**
AWS DynamoDB should be a supported storage instance.

**Describe alternatives you've considered**
There are no cloud-native storage solutions supported by Grafeas at present.

**Comment by john-tipper** [comment]
I am currently working on an implementation for this issue.

**Comment by aysylu** [comment]
Great! Looking forward to your contribution @john-tipper. Please note that going forward, we'd like to have storage bindings in separate projects, see WIP in [grafeas-pgsql](https://github.com/grafeas/grafeas-pgsql). See [Grafeas architecture slide](https://www.slideshare.net/aysylu/q-con-sp-software-supply-chain-management-with-grafeas-and-kritis/102) for more details.

**Comment by john-tipper** [comment]
Yep, sorry, should have mentioned I've seen https://github.com/grafeas/grafeas/issues/341 and will do similar.  Will be asking for creation of grafeas/grafeas-dynamodb repo when I've got something that works.

**Comment by aysylu** [comment]
@john-tipper no problem and glad to hear we're on the same page! Sounds good. I was also thinking that we can link to the dedicated storage binding projects from the Grafeas main page, but the `grafeas-dynamodb` itself doesn't have to live in the `grafeas` organization, thereby simplifying the governance and requirements specific to contributing to Grafeas. WDYT of this approach?

**Comment by john-tipper** [comment]
Hey @aysylu, I have made a PR (https://github.com/grafeas/grafeas/pull/382) that separates out implementation-specific configuration from the project configuration.  I agree with you: it'll allow for this Grafeas repo to refer in the README to other storage implementations (if you think they're good enough), but there should be no need for further changes inside this repo.  It should make #341 much easier to implement.

I've put some detailed commentary in the PR, let me know what you think - it's my first attempt at Go.

**Comment by john-tipper** [comment]
Initial commit of #380 is at https://github.com/john-tipper/grafeas-dynamodb.  It will require this PR to be merged to work correctly without manually rejigging the Go modules defined in that project.  

It provides a good example for how to use the changes made in #383 to support a new storage system and provide a server implementation that makes use of that implementation, which may well be of interest for anyone working on #341.

**Comment by john-tipper** [comment]
I think this will require a bump in version of the core `grafeas` project to `v0.1.2` to include the changes within #383.  See https://github.com/grafeas/grafeas/pull/383#issuecomment-540223357 for details.

**Comment by john-tipper** [comment]
Thanks for the release of `v0.1.2`, @aysylu.  The `grafeas` project includes the generate step in its makefile: https://github.com/grafeas/grafeas/blob/master/Makefile#L31-L39, where code is generated from protobuf files within this project.

Note that at present there is a hard-coded OS dependency coded into the generation of the dependencies: https://github.com/grafeas/grafeas/blob/master/protoc/generate.go#L7.  For me to build `grafeas` on OSX, I need to edit this file to refer to `https://github.com/google/protobuf/releases/download/v3.8.0/protoc-3.8.0-linux-x86_64.zip`.

When a dependent project (one needing `grafeas`) is built and it requires some of the protobuf dependencies, there's an error at present as they are not generated, for example:

```
build github.com/john-tipper/grafeas-dynamodb/go/v1beta1/main: cannot load github.com/grafeas/grafeas/proto/v1beta1/attestation_go_proto: cannot find module providing package github.com/grafeas/grafeas/proto/v1beta1/attestation_go_proto
```

What is the best approach to having dependent projects generate these dependencies that we import from this project?  The generation of dependencies needs to be OS-agnostic.

**Comment by john-tipper** [comment]
I've taken a look at what's happening with the `generate` step in the makefile and what is actually needed to support builds in other projects: it looks like it's just the generated source from the port files that is needed.  As a result, the issue I raised about cross-platform releases is not relevant: the only requirement for cross-platform compilation is actually this project.  It is source-code, not compiled binary libraries, that are required by dependent projects.

`go generate` is used fairly heavily in this project, which would make externalizing build logic (so code can be re-generated by dependent projects) a bit of a pain.  It also feels dirty to me, in that I think only this project should be generating code that this project defines via protobuf, rather than mandating that any other project that might want to include this project's source files needs to understand how to generate the code and must also have all necessary tooling to perform the generation available.

At present, this project does not commit the generated source code into this repo (c.f. https://github.com/grafeas/grafeas/blob/master/.gitignore), nor is the generated code contained within this project's releases (https://github.com/grafeas/grafeas/releases).  Is there a reason why we could not run `make generate` and commit the generated code to the repo, or add the generated code to the source releases?  

If we do the former (add generated code to repo), then dependent projects will be able to just run `go get` (or include this project in `go.mod`) and the dependent project should have access to all source code.  If we do the latter (add to release tarball), then dependent projects would need to download the tarball via a Makefile or `go:generate` step, which is less convenient, but still workable.

**Comment by aysylu** [comment]
@john-tipper: thanks for raising this issue and investigating. I filed #393 for the general issue that has come up before, where the Makefile is not OS-agnostic. Please feel free to chime in on the issue.

We used to commit the generated code to the repo, but cleaned it up and stopped doing this in the interests of keeping the source code clean of generated artifacts. I'm open to adding the generated code to the source releases, over checking it into the repo. Could you please confirm this would be sufficient?

**Comment by aysylu** [comment]
@john-tipper sounds good, and yes, if you could publish the solution for others to use, that'd be great! I'll release this today or latest by tomorrow and will update the thread. Thanks again for all your contributions!

**Comment by aysylu** [comment]
@john-tipper: I just released `v0.1.3` with the generated protos. PTAL and LMK if it works for you. Thanks!

**Comment by john-tipper** [comment]
@aysylu I have an initial version here: https://github.com/john-tipper/grafeas-dynamodb.  It is affected by the bug described by @judavi: #389.

To get the code to compile with the generated source code I have had to abuse the go module system and effectively bypass it: https://github.com/john-tipper/grafeas-dynamodb/blob/master/go.mod#L33.  Code from this project is downloaded via the Makefile (https://github.com/john-tipper/grafeas-dynamodb/blob/master/Makefile#L41-L44) and then used to replace the module specified.  I think this is a dirty hack, as the version specified in the module file is not necessarily that used to compile the project (it's the version is specified in the Makefile that is used).  On reflection, I think dependent projects should be able to just use the normal go module system to download all the code that they need, so I actually think the generated code from this project should be in this repo.

**Comment by aysylu** [comment]
@john-tipper thank you for looking further into this! @noahklein is looking into #389 and will update us with his findings/fixes there.

The generated protos used to be in the repo, but then we removed them in favor of keeping the API codebase cleaner. However, my understanding is that it's a common practice to check into the repo generated protos in the Go world, so I'll add them back.

Given that #389 still needs to be fixed, WDYT of releasing the code with checked in go protos as `v0.1.4-beta` (or something similar?) so you can continue making progress on your end? Once #389 is addressed, we can release the full `v0.1.4`.

**Comment by aysylu** [comment]
Update: the fix by @judavi is merged, and #398 should simplify the downstream integration with Grafeas. Once #398 is merged, I'll release a new version.

**Comment by aysylu** [comment]
`v0.1.4` is out, PTAL and LMK whether it solves the issue, @john-tipper.
